### PR TITLE
feature: Factory methods for AHS AtomArrangments by AbdullahKazi500

### LINKED ISSUE #969: Factory methods for AHS AtomArrangments
**Describe the feature you'd like**
Currently there is a single low-level constructor for [AtomArrangements](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/src/braket/ahs/atom_arrangement.py#L57), which creates an empty atom arrangement object, after which its [add](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/src/braket/ahs/atom_arrangement.py#L64) method can be used to add atoms.

Use cases of condensed matter physics often require specifying atom arrangements with regular structure, such as square, triangle, honeycomb lattices. To specify these lattices, currently one need to explicitly program the coordinates of each atom in lattice. It would be much better to be able to use pre-packaged factory methods to create these arrangements, and maybe even the most general 2-d decorated Bravais lattice too.

A good reference on decorated Bravais lattices: https://courses.physics.illinois.edu/phys560/sp2011/lecture/Phys560Notes-3.pdf

**How would this feature be used? Please describe.**
Atom arrangements could be defined by defining a regular lattice structure and a canvas where the actual sites are placed.
1. Square lattice
    ```python
    canvas_boundary_points = [(0,0), (7.5e-5, 0), (0, 7.6e-5), (7.5e-5, 7.6e-5)]
    canvas = Canvas(canvas_boundary_points)
    d = 4e-6
    register = AtomArrangement.from_square_lattice(d, canvas)
    ```
2. Rectangular lattice
     ```python
    dx = 4e-6
    dy = 6e-6
    register = AtomArrangement.from_rectangular_lattice(dx, dy, canvas)
    ```
3. Honeycomb lattice
     ```python
    d = 6e-6
    register = AtomArrangement.from_honeycomb_lattice(d, canvas)
    ```
4. Bravais lattice
     ```python
    a1 = (0, 5e-6)
    a2 = (3e-6, 4e-6)
    register = AtomArrangement.from_bravais_lattice(a1, a2, canvas)
    ```
5. Decorated Bravais lattice
     ```python
    a1 = (0, 10e-6)
    a2 = (6e-6, 8e-6)
    basis = [(0,0), (0, 5e-6), (4e-6, 2e-6)]
    register = AtomArrangement.from_bravais_lattice(a1, a2, canvas, basis=basis)
    ```

**Describe alternatives you've considered**
One alternative to the API `from_bravais_lattice` method above would be to provide `n1, n2` repetition numbers of the unit cell instead of the `Canvas` object. This would however not allow us to fit the atom arrangement into a pre-defined bounding box, and would not allow us to create staggered arrangements, where say row 1 contains unit cells 0,1,2,3, and row 2 contains unit cells 1,2,3,4, and row 3 contains unit cells 2,3,4,5.

**Additional context**
Useful links:
* https://en.wikipedia.org/wiki/Bravais_lattice
* https://courses.physics.illinois.edu/phys560/sp2011/lecture/Phys560Notes-3.pdf
* https://en.wikipedia.org/wiki/Point_in_polygon

### PR DESCRIPTION
**Pull request by AbdullahKazi500**
*Issue #, if available:*
fixes #969 
*Description of changes:*
added lattice.py polygon.py and the factory ahs all in one file 
*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by peterkomar-aws** [comment]
Hi @AbdullahKazi500 , There are good ideas here. But there is also a lot of unnecessary (and confusing) redefinitions and changes that are not pertinent to the issue at hand.

Also, to avoid future confusion, please work on this PR going forward. I will close the rest.

As a first start:
1. Please remove redefinitions
2. Please add your test to the test folder in the appropriate module.

If you need guidance how to do this, happy to point you to materials.

**Comment by peterkomar-aws** [comment]
Hi @AbdullahKazi500 , no worries. We were all there some time in our journey. Please try to follow the [contributions guidelines](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md). You can start by creating a fork of the repo, and work there on a feature branch first. Once your tests are passing locally and you are happy with the code, you can create a PR from your fork to this repo.

**Comment by peterkomar-aws** [comment]
Hi @AbdullahKazi500 , thanks for continuing the work. 

You wrote that you removed unnecessary definitions, but the [atom_arrangement.py file in your latest commit](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/05b43a96a468da13b3a0dfc48c77d477e2a2e339/src/braket/ahs/atom_arrangement.py) still contains many repeated redefinitions. E.g. the original defition of AtomArrangementItem is defined in L41, then redefined in L187, L351, and L393; AtomArrangement is originally defined in L63, then redefined in L208, L428. 

Your proposed version of the module has 527 lines, I understand that may be difficult to handle, but to move forward with this work, I ask you to review each line of change and addition that you are proposing ([here](https://github.com/amazon-braket/amazon-braket-sdk-python/pull/989/files)) and carefully consider its function and value. Once you feel that your work is ready for a careful review please click the ready for review button to flip back the PR from draft stage to open. Thank you.

**Comment by peterkomar-aws** [comment]
Hi @AbdullahKazi500 , thanks for working on this PR. 

In my previous comments, I mentioned that you had many re-definitions in your atom_arrangement.py module. This is still the case. I'd like to start a conversation with you about one particular case, `SiteType`, since you explicitly mentioned it in your message above.

The class `SiteType` is an enum class.
* It is part of the main branch, defined in [L28](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/src/braket/ahs/atom_arrangement.py#L28)
```python
class SiteType(Enum):
    VACANT = "Vacant"
    FILLED = "Filled"
```
* Its first definition appears in [L35](https://github.com/AbdullahKazi500/amazon-braket-sdk-python/blob/ead9f9709b79211091c85ccb2bc57cf78cfdd657/src/braket/ahs/atom_arrangement.py#L35) in your code, but this is already different from how it is in the main branch
```python
class SiteType(Enum):
    FILLED = "filled"
    VACANT = "vacant"
```
* **Question 1:** Why did you change the string values (filled vs Filled), and why did you change the order in which they are listed in the class definition?
* Then, a redefinition appears in [L182](https://github.com/AbdullahKazi500/amazon-braket-sdk-python/blob/ead9f9709b79211091c85ccb2bc57cf78cfdd657/src/braket/ahs/atom_arrangement.py#L182), which is identical to the definition in the main branch
```python
class SiteType(Enum):
    VACANT = "Vacant"
    FILLED = "Filled"
```
* **Question 2:** What is the purpose of this code? Why did you not leave the original definition from the main branch in place?
* Then, another redefinition appears in [L347](https://github.com/AbdullahKazi500/amazon-braket-sdk-python/blob/ead9f9709b79211091c85ccb2bc57cf78cfdd657/src/braket/ahs/atom_arrangement.py#L347), where both the attribute names and the string values are different (empty vs vacant).
```python
class SiteType(Enum):
    FILLED = "filled"
    EMPTY = "empty"
```
* **Question 3:** What is the purpose of this? `SiteType.VACANT` is a public attribute in this module; changing/eliminating it would break user code built on top of this class. There should be a very good reason to change this public API. What is your reason?
* Then, another redefinition appears in [L388](https://github.com/AbdullahKazi500/amazon-braket-sdk-python/blob/ead9f9709b79211091c85ccb2bc57cf78cfdd657/src/braket/ahs/atom_arrangement.py#L388)
```python
class SiteType(Enum):
    FILLED = "filled"
    EMPTY = "empty"
```
* **Question 4:** Why do you need this? Since this is repeating the same definition again, it seems to serve no purpose.

Since `SiteType` is such a simple class, I wish to have a productive conversation about your plans to change the atom_arrangement module through this lens.

**Comment by peterkomar-aws** [comment]
Hi @AbdullahKazi500 , this sounds like a good next step.

**Comment by peterkomar-aws** [comment]
Hi @AbdullahKazi500 , this looks very clean. Please make the edit on the [AbdullahKazi500:AbdullahKazi500-patch-8](https://github.com/AbdullahKazi500/amazon-braket-sdk-python/tree/AbdullahKazi500-patch-8) branch. Then I can provide line-by-line review as part of this PR.

**Comment by peterkomar-aws** [comment]
Hi @AbdullahKazi500 , thanks for adding the changes to the AHS.py file. Since you contribution will be modifications to the atom_arrangement.py file, please add the changes there. If you replace the content of that file with what you have in you AHS.py, then I can do a more through review, beacause I will see the changes with respect to the original atom_arrangement.py file. Thanks.

File: AHS.py
Comment (peterkomar-aws) [inline AHS.py:1]: Please remove this file, and make all edits to the atom_arrangement.py module and the test_atom_arrangement.py module.

File: src/braket/ahs/atom_arrangement.py [hunk 1]
```diff
@@ -23,18 +23,24 @@
 import numpy as np
 
 from braket.ahs.discretization_types import DiscretizationError, DiscretizationProperties
+from typing import Union, Tuple, List, Iterator
```
Comment (peterkomar-aws) [inline src/braket/ahs/atom_arrangement.py:26]: Please merge this line with L21 above

File: src/braket/ahs/atom_arrangement.py [hunk 2]
```diff
@@ -23,18 +23,24 @@
...
+from dataclasses import dataclass
```
Comment (peterkomar-aws) [inline src/braket/ahs/atom_arrangement.py:27]: This is a repeated import (same as L17), please remove.

File: src/braket/ahs/atom_arrangement.py [hunk 3]
```diff
@@ -23,18 +23,24 @@
...
+import numpy as np
```
Comment (peterkomar-aws) [inline src/braket/ahs/atom_arrangement.py:28]: This is a repeated import (same as line 23), please remove.

File: src/braket/ahs/atom_arrangement.py [hunk 4]
```diff
@@ -23,18 +23,24 @@
...
+from decimal import Decimal
```
Comment (peterkomar-aws) [inline src/braket/ahs/atom_arrangement.py:29]: This is a repeated import (same as line 18), please remove.

File: src/braket/ahs/atom_arrangement.py [hunk 5]
```diff
@@ -23,18 +23,24 @@
...
+from enum import Enum
```
Comment (peterkomar-aws) [inline src/braket/ahs/atom_arrangement.py:30]: This is a repeated import (same as line 19), please remove.

File: src/braket/ahs/atom_arrangement.py [hunk 6]
```diff
@@ -23,18 +23,24 @@
...
+from numbers import Number
```
Comment (peterkomar-aws) [inline src/braket/ahs/atom_arrangement.py:31]: This is a repeated import (same as line 20), please remove.

File: src/braket/ahs/atom_arrangement.py [hunk 7]
```diff
@@ -23,18 +23,24 @@
...
 
 
+# Define SiteType enum
```
Comment (peterkomar-aws) [inline src/braket/ahs/atom_arrangement.py:34]: This comment does not tell anything beyond what the next line is telling already. Please remove.

File: src/braket/ahs/atom_arrangement.py [hunk 8]
```diff
@@ -23,18 +23,24 @@
...
 class SiteType(Enum):
-    VACANT = "Vacant"
```
Comment (peterkomar-aws) [inline src/braket/ahs/atom_arrangement.py:29]: Let's leave `SiteType` alone. The original definition works. There is nothing to fix here.

File: src/braket/ahs/atom_arrangement.py [hunk 9]
```diff
@@ -63,56 +69,153 @@ def __init__(self):
 
     def add(
         self,
-        coordinate: Union[tuple[Number, Number], np.ndarray],
+        coordinate: Union[Tuple[Number, Number], np.ndarray],
         site_type: SiteType = SiteType.FILLED,
-    ) -> AtomArrangement:
+    ) -> "AtomArrangement":
```
Comment (peterkomar-aws) [inline src/braket/ahs/atom_arrangement.py:74]: The original hint referred to the `AtomArrangement` class as a return type. What does turning it into a string literal achieve here?

File: src/braket/ahs/atom_arrangement.py [hunk 10]
```diff
@@ -63,56 +69,153 @@ def __init__(self):
...
         """Add a coordinate to the atom arrangement.
-
```
Comment (peterkomar-aws) [inline src/braket/ahs/atom_arrangement.py:70]: We are using Google formatted docstrings in this repo (see more here: https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings), please don't remove these empty lines.

Comment (peterkomar-aws) [reply src/braket/ahs/atom_arrangement.py:70]: Same for many other removed lines in the docstrings below.

File: src/braket/ahs/atom_arrangement.py [hunk 11]
```diff
@@ -63,56 +69,153 @@ def __init__(self):
...
+            for site in self._sites:
+                new_coordinates = tuple(
+                    round(Decimal(c) / position_res) * position_res for c in site.coordinate
+                )
+                discretized_arrangement.add(new_coordinates, site.site_type)
+            return discretized_arrangement
+        except Exception as e:
+            raise DiscretizationError(f"Failed to discretize register {e}") from e
+
+# Updated
```
Comment (peterkomar-aws) [inline src/braket/ahs/atom_arrangement.py:127]: I see you appended the latest version of the file starting here. Instead, please replace the content of the entire file with this, so I can do a more careful, line-by-line review. Thanks.

**Review by peterkomar-aws — CHANGES_REQUESTED** [review]
I see you appended the latest version of the file. Instead, please replace the content of the entire file with this, so I can do a more careful, line-by-line review. Thanks.

**Review by peterkomar-aws — CHANGES_REQUESTED** [review]
Great start. You have the right structure for the factory methods. 

The main thing that needs change is the implementation of the `from_decorated_bravais_lattice` method. Currently it cuts off too many sites, sites which would fit on the canvas. See my comment for line 141. This is not easy to solve. Looking forward to your idea how to make this method work for arbitrary vec_a and vec_b values. 

The goal is the fill up the entire canvas, and only remove sites that would truly fall outside of the canvas boundary.

**Comment by peterkomar-aws** [comment]
Hi @AbdullahKazi500 , can you explain the logic behind `lattice_x_max = min(x_max, x_max - vec_a[0] + vec_b[0])`? (It is very suspicious that vec_a and vec_b are not treated symmetrically by this expression. Logically, they are interchangeable.)

**Comment by peterkomar-aws** [comment]
Hi @AbdullahKazi500 , thanks for your contributions. Looking forward to your contributions to this work and repo. Thank you. (I mark the PR draft again; feel free to add commits to it an open it again if you think it's ready for another round of reviews. Thanks.)

**Comment by peterkomar-aws** [comment]
Hi @AbdullahKazi500 , the implementation you proposed in your [previous comment](https://github.com/amazon-braket/amazon-braket-sdk-python/pull/989#issuecomment-2181540249),  

```python
@classmethod
def from_decorated_bravais_lattice(cls, lattice_vectors: List[Tuple[float, float]], decoration_points: List[Tuple[float, float]], canvas_boundary_points: List[Tuple[float, float]]) -> AtomArrangement:
    """Create an atom arrangement with a decorated Bravais lattice."""
    arrangement = cls()
    vec_a, vec_b = np.array(lattice_vectors[0]), np.array(lattice_vectors[1])
    x_min, y_min = canvas_boundary_points[0]
    x_max, y_max = canvas_boundary_points[2]

    # Iterate over lattice points within the canvas boundaries
    i = 0
    while True:
        origin = i * vec_a
        if origin[0] > x_max:
            break
        j = 0
        while True:
            point = origin + j * vec_b
            if point[1] > y_max:
                break
            if x_min <= point[0] <= x_max and y_min <= point[1] <= y_max:
                for dp in decoration_points:
                    decorated_point = point + np.array(dp)
                    if x_min <= decorated_point[0] <= x_max and y_min <= decorated_point[1] <= y_max:
                        arrangement.add(tuple(decorated_point))
            j += 1
        i += 1

    return arrangement
```
has the same issue which [I pointed out previously](https://github.com/amazon-braket/amazon-braket-sdk-python/pull/989#discussion_r1635202066), namely 

> To fill in the entire canvas, depending on the values of vec_a and vec_b, it may not be enough to start from i=0 and j=0. E.g. if vec_a = (1,1) and vec_b = (-1, 1), your current implementation leaves a large empty triangle at the lower right side on a rectangular canvas.

More specifically, running the following test
```python
lattice_vectors = (
    (1, 1),
    (-1, 1),
)
decoration_points = (
    (0, 0),
)
canvas_boundary_points = (
    (0, 0),
    (10, 0),
    (10, 10),
    (0, 10),
)

register = AtomArrangement.from_decorated_bravais_lattice(lattice_vectors, decoration_points, canvas_boundary_points)
```
produces a `register` which looks like this
Screenshot 2024-06-24 at 10 46 12 AM

**Comment by peterkomar-aws** [comment]
Hi @AbdullahKazi500 , regarding these lines in your proposal,
```python
n_steps_x = int(np.ceil((x_max - x_min) / np.linalg.norm(vec_a)))
n_steps_y = int(np.ceil((y_max - y_min) / np.linalg.norm(vec_b)))
````
can you explain why `n_steps_x` is computed only from `x_max` and `x_min`, while you are using it to set a hard limit on how many times you step with `vec_a` later? See here
```python
for i in range(-n_steps_x, n_steps_x + 1):
            for j in range(-n_steps_y, n_steps_y + 1):
                origin = i * vec_a + j * vec_b
```
and why `n_step_y` is computed only from `y_max` and `y_min`, while you are using it to set a hard limit on how many times you step step with `vec_b`?

Any implementation that does not treat `vec_a` and `vec_b` symmetrically is very suspicious, because if I reorder the `lattice_vectors` input list (which effectively swaps the values of `vec_a` and `vec_b`), there is a good chance the returned AtomArrangement will be different.

**Comment by peterkomar-aws** [comment]
Hi @AbdullahKazi500 , I appreciate that this implementation is symmetric in its treatment of `vec_a` and `vec_b`, this is a good first step. The implementation you are proposing in your comment above, namely,
```python
@classmethod
def from_decorated_bravais_lattice(cls, lattice_vectors: List[Tuple[float, float]], decoration_points: List[Tuple[float, float]], canvas_boundary_points: List[Tuple[float, float]]) -> AtomArrangement:
    """Create an atom arrangement with a decorated Bravais lattice."""
    if len(lattice_vectors) != 2 or any(len(vec) != 2 for vec in lattice_vectors):
        raise ValueError("Invalid lattice vectors provided.")
    if any(len(dp) != 2 for dp in decoration_points):
        raise ValueError("Invalid decoration points provided.")
    if len(canvas_boundary_points) != 4:
        raise ValueError("Invalid canvas boundary points provided.")

    arrangement = cls()
    vec_a, vec_b = np.array(lattice_vectors[0]), np.array(lattice_vectors[1])
    x_min, y_min = canvas_boundary_points[0]
    x_max, y_max = canvas_boundary_points[2]

    def is_within_canvas(point: np.ndarray) -> bool:
        return x_min <= point[0] <= x_max and y_min <= point[1] <= y_max

    # Determine the step directions and limits symmetrically
    n_steps_x = int(np.ceil(max(x_max - x_min, y_max - y_min) / np.linalg.norm(vec_a)))
    n_steps_y = int(np.ceil(max(x_max - x_min, y_max - y_min) / np.linalg.norm(vec_b)))

    for i in range(-n_steps_x, n_steps_x + 1):
        for j in range(-n_steps_y, n_steps_y + 1):
            origin = i * vec_a + j * vec_b
            for dp in decoration_points:
                decorated_point = origin + np.array(dp)
                if is_within_canvas(decorated_point):
                    arrangement.add(tuple(decorated_point))

    return arrangement
```
also does not properly fill out the canvas for the test I proposed above, i.e.
```python
lattice_vectors = (
    (1, 1),
    (-1, 1),
)
decoration_points = (
    (0, 0),
)
canvas_boundary_points = (
    (0, 0),
    (10, 0),
    (10, 10),
    (0, 10),
)

register = AtomArrangement.from_decorated_bravais_lattice(lattice_vectors, decoration_points, canvas_boundary_points)
```
produces a `register` that looks like this
Screenshot 2024-06-28 at 10 44 11 AM

I suspect this is due to the premature stopping of the `for i` and `for j` loops, i.e. `n_steps_x` and `n_steps_y` are too small. My guess is this is due to the fact that you are computing them by dividing the Euclidean norm of `vec_a` and `vec_b`, which is generally larger than the projection of the said vectors in the x and y directions, respectively.

**Comment by rmshaffer** [comment]
Closing due to inactivity. If you'd like to continue working on this, feel free to re-open!

**Pull request closed without merging**